util/reggen/uvm_reg_base.sv.tpl

```diff
@@ -261,9 +266,34 @@ endpackage
 %>\
   class ${class_name} extends ${dv_base_prefix}_reg;
     // fields
-% for f in reg.fields:
+<%
+  suffix = ""
+  start_idx = 0
+  compact_field_inst_name = ""
+  if isinstance(mr, MultiRegister):
+    if mr.compact and len(mr.reg.fields) == 1:
+      compact_field_inst_name = mr.reg.fields[0].name.lower()
+      fields = mr.regs[reg_idx].fields
+      if len(fields) > 1:
+        num_fields_per_reg = 32 // fields[0].bits.width()
+        start_idx = num_fields_per_reg * reg_idx
+        end_idx = start_idx + len(fields) - 1
+        if start_idx > 0:
+          suffix = f'[{start_idx}:{end_idx}]'
+        else:
+          suffix = f'[{end_idx+1}]'
+    else:
+      fields = mr.reg.fields
+  else:
+    fields = reg.fields
```

Comment:
This took me a *long* time to understand but I think there are some simple tweaks to make it easier to read. How about the following? I've slightly re-ordered things and have added some comments and assertions explaining what's going on.
```suggestion
  if mr is None:
    fields = reg.fields
  else:
    if not mr.compact:
      fields = mr.reg.fields
    else:
      fields = mr.regs[reg_idx].fields
      compact_field_inst_name = mr.reg.fields[0].name.lower()
      if mr.dv_compact:
        # The dv_compact flag means that the fields of the multi-reg divide equally into registers.
        # In this case, there's an array of registers and make_ral_pkg_reg_class() gets called once
        # to define that array's type, using the fields of the first register in the replication.
        assert reg_idx == 0
        if len(fields) > 1:
          suffix = f'[{len(fields)}]'
      else:
        # In this case, the multi-register is "compact", so there might be multiple copies of its
        # single field in each generated register. But dv_compact is false, which probably means
        # that the fields didn't divide equally into a whole number of registers. In this case, we
        # are generating a different class for each output register and should spit out fields
        # accordingly. Note that we generate an array, even if len(fields) = 1. If that happens, we
        # know we're on the last generated register, so want to keep everything uniform.
        num_fields_per_reg = 32 // fields[0].bits.width()
        start_idx = num_fields_per_reg * reg_idx
        end_idx = start_idx + len(fields) - 1
        suffix = f'[{start_idx}:{end_idx}]'
```
Note that this changes the generated code for register 0 of a non-dv-compact replicated register. The point is that we probably don't want to treat it differently from the other registers.

Reply:
thanks, this is more readable.
adjust the suffix to avoid style check errors, as var[0:N] isn't allowed.
```
        if start_idx == 0:
          suffix = f'[{len(fields)}]'
        else:
          end_idx = start_idx + len(fields) - 1
          suffix = f'[{start_idx}:{end_idx}]'
```

Reply:
Hmm. That seems a weird lint rule. Maybe the best thing is to do what I suggested and switch to `[N:0]` instead? That'll allow things to stay consistent and I'm pretty certain we don't have any lint rules disallowing that :-)

Reply:
we do disallow little-endian order for an unpacked array and suggest to use `a[N]` instead of `a[0:N-1]`

https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#unpacked-ordering

Reply:
Well, I think that we should either use `[0:9]` and `[10:19]` or use `[10]` for both. Mixing the two seems really silly to me (style guide or no...)

Reply:
Thanks @weicaiyang for the work you're putting into this and trying to find a middle ground between various slightly conflicting requirements.

I agree with @rswarbrick, it would be great if we can keep the syntax consistent and therefore our template simple(r).

Unless we have reasons otherwise, I'd just go with `[N]` consistently as "path of least resistance": it's what the style guide suggests and it works in lint. There might be of course reasons that I'm not aware of right now that suggest something like `[10:19]` is preferable. In that case, let's have the discussion and change the style guide (and our lint setup).

Reply:
The `[10:19]` thing was a cunning idea from @sriyerg. The idea is that then you can do things like `my_reg[1].my_field[10]` rather than having to calculate that field 10 is at index zero of the first register.

Reply:
I think this style requirement is very reasonable, otherwise, we will see all kinds of unpacked array, like `array[N], array[0:N-1], array[N-1:0]`.
Since we have a reason of using this `[0:N-1]` here, let me add a inline waiver for it